### PR TITLE
Enabling custom attributes to be passed through SAML assertions.

### PR DESCRIPTION
**Pull request by msisskin**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java [hunk 1]
```diff
@@ -133,7 +151,42 @@ public String getDecision(String subject, String resource, String action, String
...
 
+        if (!appToPDPClientMap.containsKey(appId)) {
+            throw new EntitlementProxyException("Invalid App Id");
+        }
+        Attribute subjectAttribute = new Attribute(
+                URN_OASIS_NAMES_TC_XACML_1_0_SUBJECT_CATEGORY_ACCESS_SUBJECT, URN_OASIS_NAMES_TC_XACML_1_0_SUBJECT_SUBJECT_ID, ProxyConstants.DEFAULT_DATA_TYPE, subject);
+        Attribute actionAttribute = new Attribute(URN_OASIS_NAMES_TC_XACML_3_0_ATTRIBUTE_CATEGORY_ACTION, URN_OASIS_NAMES_TC_XACML_1_0_ACTION_ACTION_ID, ProxyConstants.DEFAULT_DATA_TYPE, action);
+        Attribute resourceAttribute = new Attribute(URN_OASIS_NAMES_TC_XACML_3_0_ATTRIBUTE_CATEGORY_RESOURCE, URN_OASIS_NAMES_TC_XACML_1_0_RESOURCE_RESOURCE_ID, ProxyConstants.DEFAULT_DATA_TYPE, resource);
+        Attribute environmentAttribute = new Attribute(URN_OASIS_NAMES_TC_XACML_3_0_ATTRIBUTE_CATEGORY_ENVIRONMENT, URN_OASIS_NAMES_TC_XACML_1_0_ENVIRONMENT_ENVIRONMENT_ID, ProxyConstants.DEFAULT_DATA_TYPE, environment);
+        Attribute[] tempArr = {actionAttribute, resourceAttribute, environmentAttribute};
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java]: from line 172 to 176, exceeds the maximum width

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/util/CarbonEntityResolver.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import java.io.IOException;
+
+public class CarbonEntityResolver implements EntityResolver {
+
+    @Override
+    public InputSource resolveEntity(String publicId, String systemId) throws SAXException,
+            IOException {
+        throw new SAXException("SAML request contains invalid elements. Possible XML External Entity (XXE) attack.");
+    }
+
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/util/CarbonEntityResolver.java]: extra new line

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/util/CarbonEntityResolver.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2014, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/util/CarbonEntityResolver.java]: change the license header

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxyFactory.java [hunk 4]
```diff
@@ -79,7 +79,12 @@ private PEPProxyFactory(){
                 if (!serverUrl.endsWith("/")) {
                     serverUrl += "/";
                 }
-
+                if (appConfig.get(USER_NAME) == null || appConfig.get(USER_NAME).length() == 0) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxyFactory.java]: possible NPE in appConfig. If appId is not in appToPDPClientConfigMap, appConfig can be null

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java [hunk 5]
```diff
@@ -133,7 +151,42 @@ public String getDecision(String subject, String resource, String action, String
...
+     * @param subject     XACML 2.0 subject
+     * @param attributes  An array of XACML attributes to send to the PDP Proxy
+     * @param resource    XACML 2.0 resource
+     * @param action      XACML 2.0 action
+     * @param environment XACML 2.0 environments
+     * @param appId       specific appID in the PDP Proxy, there can be many PDPs configured for appID. Each App can have distinct PDPs
+     * @return the Entitlement Decision as a String
+     * @throws Exception
+     */
+    public String getDecision(String subject, Attribute[] attributes, String resource, String action, String environment, String appId) throws Exception {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java]: formatting

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java [hunk 6]
```diff
@@ -109,6 +110,23 @@ public String getDecision(Attribute[] attributes, String appId) throws Exception
...
+     *
+     * @param subject     XACML 2.0 subject
+     * @param attributes  XACML 2.0 custom attributes
+     * @param resource    XACML 2.0 resource
+     * @param action      XACML 2.0 action
+     * @param environment XACML 2.0 environments
+     * @return the Entitlement Decision as a String
+     * @throws Exception
+     */
+    public String getDecision(String subject, Attribute[] attributes, String resource, String action, String environment) throws Exception {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java]: formatting

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java [hunk 7]
```diff
@@ -133,7 +151,42 @@ public String getDecision(String subject, String resource, String action, String
         Attribute[] tempArr = {subjectAttribute, actionAttribute, resourceAttribute, environmentAttribute};
         return getDecision(tempArr, appId);
     }
+    
+    /**
+     * This method is used to get the Entitlement decision for the provided subject,attributes,resource,action and environment using the provided appID of the PDP defaultProxy
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxy.java]: formatting

File: components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxyFactory.java [hunk 8]
```diff
@@ -79,7 +79,12 @@ private PEPProxyFactory(){
...
+                    throw new EntitlementProxyException("userName cannot be null or empty");
+                }
+                if (appConfig.get(PASSWORD) == null || appConfig.get(PASSWORD).length() == 0) {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.entitlement.proxy/src/main/org/wso2/carbon/identity/entitlement/proxy/PEPProxyFactory.java]: possible NPE

**Comment by piraveena** [comment]
@msisskin Thanks for your contribution. We can see some conflicts need to be resolved. Please fix the comments and resolve the conflicts

**Comment by emswbandara** [comment]
Closing since comments and conflicts are not addressed.

**Pull request closed without merging**